### PR TITLE
implement a middleware named replicate

### PR DESCRIPTION
**Pull request by MikhailPopov**
I implemented a middleware that reads requests and responses and sends them to a message broker. The middleware doesn't affect requests and responses.

### CODE REVIEW

File: pkg/server/middleware/middlewares.go [hunk 1]
```diff
@@ -339,6 +340,20 @@ func (b *Builder) buildConstructor(ctx context.Context, middlewareName string) (
...
 	}
 
+	// Replicate
+	if config.Replicate != nil {
+		middleware = func(next http.Handler) (http.Handler, error) {
+			producer, err := replicate.NewKafkaPublisher(config.Replicate.Topic, config.Replicate.Brokers)
+			// todo find the correct way to close the producer (producer.Close())
+			if err != nil {
+				noOpHandler := http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {})
+				err = errors.New(strings.Join([]string{"Replicate: failed to create a producer", err.Error()}, ": "))
```
Comment (raproid) [inline pkg/server/middleware/middlewares.go:350]: See, the author is using the article "a" here. It's a good habit to include them in your message content.

File: pkg/middlewares/replicate/replicate_test.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		require.NoError(t, err)
+
+		request := httptest.NewRequest(method, URL, strings.NewReader(expectedBody))
+		recorder := httptest.NewRecorder()
+		replicate.ServeHTTP(recorder, request)
+
+		assert.Equal(t, http.StatusOK, recorder.Code, "status code was changed by middleware")
+		assert.Equal(t, expectedBody, recorder.Body.String(), "response body was changed by middleware")
+		assert.Len(t, recorder.Header(), 2, "length of headers was changed by middleware")
+		assert.Equal(t, recorder.Header().Get(expectedHeader), expectedValue, "header was changed by middleware")
```
Comment (raproid) [inline pkg/middlewares/replicate/replicate_test.go]: "...by the middleware"

Comment (raproid) [reply pkg/middlewares/replicate/replicate_test.go]: the same for other messages containing "by middleware" in this file

File: pkg/middlewares/replicate/producer_test.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+			name:    "Fails with nil brokers",
+			topic:   "test-topic",
+			brokers: nil,
+			expMsg:  "at least one broker is required",
+		},
+		{
+			name:    "Fails to connect broker",
+			topic:   "test-topic",
+			brokers: []string{"broker:1234"},
+			expMsg:  "cannot create Kafka producer",
```
Comment (raproid) [inline pkg/middlewares/replicate/producer_test.go:38]: "a Kafka producer"

File: pkg/middlewares/replicate/replicate.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+// replicate is a middleware used to send copies of requests and responses to arbitrary service
+type replicate struct {
+	next     http.Handler
+	name     string
+	producer Producer
+}
+
+// New creates a new http handler.
+func New(ctx context.Context, next http.Handler, producer Producer, name string) (http.Handler, error) {
+	log.FromContext(middlewares.GetLoggerCtx(ctx, name, typeName)).Debug("Creating middleware")
```
Comment (raproid) [inline pkg/middlewares/replicate/replicate.go:26]: Are you creating an instance here or am I mistaken? "Creating a middleware instance"?

File: pkg/middlewares/replicate/replicate_test.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNewReplicate(t *testing.T) {
+	t.Run("Creates instance with valid params", func(t *testing.T) {
```
Comment (raproid) [inline pkg/middlewares/replicate/replicate_test.go]: "Creates an instance..."

File: pkg/middlewares/replicate/replicate_test.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+		name := "test-replicate"
+
+		handler, err := New(context.Background(), next, producer, name)
+		assert.NoError(t, err)
+		assert.NotNil(t, handler)
+	})
+}
+
+func TestReplicate(t *testing.T) {
+	t.Run("Don't affect on request and response", func(t *testing.T) {
```
Comment (raproid) [inline pkg/middlewares/replicate/replicate_test.go]: What are you trying to say here? Can you explain in a little bit more detail?

File: pkg/middlewares/replicate/replicate.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+	"github.com/traefik/traefik/v2/pkg/log"
+	"github.com/traefik/traefik/v2/pkg/middlewares"
+)
+
+const (
+	typeName = "Replicate"
+)
+
+// replicate is a middleware used to send copies of requests and responses to arbitrary service
```
Comment (raproid) [inline pkg/middlewares/replicate/replicate.go]: "...to an arbitrary service"

File: pkg/middlewares/replicate/replicate_test.go (hunk 6 above)
Comment (raproid) [reply pkg/middlewares/replicate/replicate_test.go]: fixed

**Review by raproid — APPROVED** [review]

**Review by pavlushev — APPROVED** [review]

**Pull request merged**